### PR TITLE
feat: add find api

### PR DESCRIPTION
**Pull request by manucorporat**

### CODE REVIEW

File: bblfsh/__main__.py [hunk 1]
```diff
@@ -9,26 +9,55 @@ def setup():
     parser = argparse.ArgumentParser(
         description="Query for a UAST to Babelfish and dump it to stdout."
     )
+    parser.add_argument('filename', type=str,
```
Comment (juanjux) [inline bblfsh/__main__.py]: This change seems to be unrelated with the usage of `libuast`. Was it requested by the machine learning team?

File: bblfsh/client.py [hunk 2]
```diff
@@ -55,8 +56,12 @@ def parse(self, filename, language=None, contents=None, timeout=None,
         request = ParseRequest(filename=os.path.basename(filename),
                                    content=contents,
                                    language=self._scramble_language(language))
-        response = self._stub.Parse(request, timeout=timeout)
-        return response
+        return self._stub.Parse(request, timeout=timeout)
+
+    def parse_uast(self, filename, language=None, contents=None, timeout=None,
```
Comment (juanjux) [inline bblfsh/client.py]: This should be `def parse(` as in `~master`.

File: bblfsh/__main__.py [hunk 3]
```diff
@@ -9,26 +9,55 @@ def setup():
...
 
 def main():
     args = setup()
     if not args.disable_bblfsh_autorun:
         ensure_bblfsh_is_running()
-    client = BblfshClient(args.endpoint)
-    print(client.parse(args.file, args.language))
 
+    client = BblfshClient(args.endpoint)
+    root = client.parse_uast(args.filename, args.language)
```
Comment (juanjux) [inline bblfsh/__main__.py]: `root = client.parse(`...

File: bblfsh/test.py [hunk 4]
```diff
@@ -8,6 +8,48 @@
...
+
+        node3 = NodeNative("block")
+        node4 = NodeNative("method")
+        node5 = NodeNative("identifier")
+        node5.token = "second"
+        node6 = NodeNative("block")
+        node8 = NodeNative("loop")
+
+        root.add_child(node1)
+
```
Comment (juanjux) [inline bblfsh/test.py]: Spurious whitespace?, probably all the add_child statements would be better on a single logica paragraph (without whitespace between them).

File: Makefile [hunk 5]
```diff
@@ -2,18 +2,26 @@ PYTHON ?= python3
...
+	bblfsh/github/com/bblfsh/sdk/__init__.py \
+	bblfsh/github/com/bblfsh/sdk/uast/__init__.py \
+	bblfsh/github/com/bblfsh/sdk/protocol/__init__.py
+
+libuast:
+	git clone git@github.com:manucorporat/libuast.git libuast
+	cd libuast && cmake . && make && make install
+
+bblfsh/pyuast.py: libuast
+	cd bblfsh && swig -c++ -python -I/usr/local/include uast.i
```
Comment (juanjux) [inline Makefile]: So `swig` is needed, at least for `make`. Since the `.i` and `pyuast.py` files are included, couldn't we remove the `all:` target dependency on this one? (but keeping this in the Makefile just in case we want to regenerate `pyuast.py`).

File: bblfsh/pyuast.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+static PyObject *py_find(PyObject *self, PyObject *args)
+{
+  PyObject *obj = NULL;
+  const char *query = NULL;
+  if (!PyArg_ParseTuple(args, "Os", &obj, &query))
+  {
+    return NULL;
+  }
+  printf("BEFORE FIND!! %p\n", obj);
```
Comment (juanjux) [inline bblfsh/pyuast.cpp]: These things usually go inside an `#ifdef DEBUG`

**Comment by juanjux** [comment]
Travis still failing ^^^^^^^

File: bblfsh/__main__.py [hunk 7]
```diff
@@ -1,6 +1,8 @@
 import argparse
 import sys
 
+from bblfsh._pyuast import find
```
Comment (abeaumont) [inline bblfsh/__main__.py]: I don't think pyuast should be a private module, should it?

File: bblfsh/client.py [hunk 8]
```diff
@@ -24,13 +24,14 @@ def __init__(self, endpoint):
                          for example "0.0.0.0:9432"
         :type endpoint: str
         """
+        Node.initialize()
```
Comment (abeaumont) [inline bblfsh/client.py]: I don't think a libuast `Node` should be initialized in a `GRPC` client, why is it this way?

File: bblfsh/pyuast.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    NULL};
+
+PyMODINIT_FUNC PyInit__pyuast(void)
+{
+  return PyModule_Create(&moduledef);
+}
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (abeaumont) [inline bblfsh/pyuast.cpp]: Newline missing

File: bblfsh/__main__.py (hunk 7 above)
Comment (juanjux) [inline bblfsh/__main__.py]: I don't know if there is specified in PEP8 but there is an old convention in Python of using a leading underline for binary modules.

File: bblfsh/pyuast.c [hunk 10]
```diff
@@ -0,0 +1,119 @@
+#include "uast.h"
```
Comment (abeaumont) [inline bblfsh/pyuast.c]: Is there a code style for python modules? If not, I guess we should follow Google C++ code style, which is not been followed here.

File: bblfsh/pyuast.c [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+#include <stdint.h>
+#include <Python.h>
+
+static const char *read_str(const void *data, const char *prop)
+{
+  PyObject *node = (PyObject *)data;
+  PyObject *attribute = PyObject_GetAttrString(node, prop);
+  const char *a = PyUnicode_AsUTF8(attribute);
```
Comment (abeaumont) [inline bblfsh/pyuast.c]: `return PyUnicode_AsUTF8(attribute)`

File: bblfsh/pyuast.c [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+      .token = get_token,
+      .children_size = get_children_size,
+      .children = get_child,
+      .roles_size = get_roles_size,
+      .roles = get_role,
+  });
+  return PyModule_Create(&moduledef);
+}
+
+#ifdef __cplusplus
```
Comment (abeaumont) [inline bblfsh/pyuast.c]: This should be removed

File: setup.py [hunk 13]
```diff
@@ -1,5 +1,27 @@
-from setuptools import setup, find_packages
+import sys
+import subprocess
+from setuptools import setup, find_packages, Extension
 
+LIBRARIES = ['xml2']
+SOURCES = ['bblfsh/pyuast.c']
```
Comment (abeaumont) [inline setup.py]: These are not constants, they shouldn't be upcase

File: bblfsh/pyuast.c (hunk 10 above)
Comment (juanjux) [reply bblfsh/pyuast.c]: Not that I know for - or resulted from a quick search. Since @manucorporat probably won't be working on this anymore I'll change it soon.

File: setup.py (hunk 13 above)
Comment (juanjux) [reply setup.py]: I've seen some uppercased module-globals on Python that were not constants but it doesn't seem to be a consensus, PEP8 says that constants are uppercase but doesn't say anything about module (mutable) globals but I think that making them lowercase avoids confusion, yes.

**Comment by juanjux** [comment]
Closing it because I'll reopen this PR from my branch - I already implemented all the suggested changes, but found a crashing bug with some parameters for `find()`, debugging it right now.

**Comment by juanjux** [comment]
To have both PRs linked, the new (unfinished) one is bblfsh/client-python/pull/23
